Execute EnrichPolicyRunner on a non dedicated master node.

martijnvg opened the pull request:
In case a cluster consists out of multiple nodes then delegate the execution of EnrichPolicyRunner
to a node that is not the elected master and nodes that are not dedicated master.

Closes #70436

martijnvg committed:
Execute EnrichPolicyRunner on a non dedicated master node.

Introduce an internal action that the execute policy action delegates to.
This to ensure that the actual policy execution is never executed on the elected master node
or dedicated master nodes. In case the cluster consists out of a single node then
the internal action will attempt to execute on the current/local node.

The actual enrich policy execution is encapsulated in the `EnrichPolicyRunner` class.
This class manages the execution of several API calls, so this itself isn't doing anything heavy.
However the coordination of these api calls (in particular the reindex api call) may involve
some non-neglectable work/overhead and this shouldn't be performed on the elected master
or any other dedicated master node.

Closes #70436

elasticmachine commented:
Pinging @elastic/es-core-features (Team:Core/Features)

danhermann reviewed (approved):
General structure and test coverage LGTM. I'm less familiar with the nitty-gritty of tasks but would be interested in more info on that at some point.

martijnvg committed:
fix nit

Co-authored-by: Dan Hermann <danhermann@users.noreply.github.com>

jbaiera reviewed (commented):
Some questions regarding order of locking/unlocking policies. I like the idea here, but I think the policy locking/unlocking is tricky, especially with `waitForComplete=false`

martijnvg committed:
Merge remote-tracking branch 'es/master' into enrich_delegate_policy_runner_to_non_dedicated_master_node

martijnvg commented on x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/action/InternalExecutePolicyAction.java:
```diff
@@ -0,0 +1,170 @@
+
+                if (request.isWaitForCompletion() == false) {
+                    TaskId taskId = new TaskId(clusterState.nodes().getLocalNodeId(), task.getId());
+                    actionListener.onResponse(new Response(taskId));
```
Good catch, I missed this. Yes, it does unlock the policy on the master node, which shouldn't happen.
The policy should only be unlocked here if wait_for_completing is set to `true`.

I think in the case if `wait_for_completion` is set to `false` then the response should be returned without unlocking the policy.  At the same time the get task action should be invoked with
that task id with `wait_for_completion=true` and if that returns then the policy should be unlocked.

martijnvg replied:
What do you think about something like this: c8822544aa758abb404414e85468d58abc2c3d49 ?

jbaiera replied:
That certainly could work!

A second thought: Since the master would have a listener waiting on the task to complete anyway, is there any compelling reason that the master node shouldn't handle the `wait_for_completion` feature itself? Making the internal policy action always wait for completion might simplify the code.

There's a bit more load put on the master in that second case (creating the task and cleaning it up), but that seems quite small compared to orchestrating the entire policy execution.

martijnvg replied:
> Since the master would have a listener waiting on the task to complete anyway, is there any compelling reason that the master node shouldn't handle the wait_for_completion feature itself? 

I think this is not possible? The master node wouldn't know about the task id of the policy execution on the node that the policy gets executed on. Only the node where the policy execution actually happens knows this and can return that to the master node. Otherwise we can't report the task id of the actual policy execution.

jbaiera replied:
This makes sense to me. Waiting on the task result sounds like a good solution to me then!

martijnvg commented on x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/action/InternalExecutePolicyAction.java:
```diff
@@ -0,0 +1,170 @@
+        protected void doExecute(Task transportTask, Request request, ActionListener<Response> actionListener) {
+            var clusterState = clusterService.state();
+            var node = selectNodeForPolicyExecution(clusterState.nodes());
+            if (clusterState.nodes().getLocalNode().equals(node) == false) {
+                var handler = new ActionListenerResponseHandler<>(actionListener, Response::new);
+                transportService.sendRequest(node, NAME, request, handler);
+                return;
+            }
```
Yes, your analysis is right here. It can do more hops than it should do.  I think this can be addressed by changing `selectNodeForPolicyExecution(...)` method the return local node if current node is not elected master node. This should work, since the first time this internal action is invoked is always on the elected master node. The second time this logic is executed on another node then checks checks prevents this logic from selecting yet another node.

Or we can make it more explicit by adding a node id field (of type string) to the request class of this internal action. This would be set when this internal action is executed on the elected master node then on the target node it will always execute (having a check something like this: `if (clusterState.nodes().getLocalNode().id().equals(request.targetNodeId()) {`.

jbaiera replied:
I think the core of the issue here (We don't want this action running on master) is solved simplest by that first suggestion. As long as the preconditions for which node it is on are met then it can just execute as normal.

martijnvg replied:
Adjusted the logic: f25f1e07d8305b246bc1f55a930e30aff29c09d2

martijnvg commented on x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/action/InternalExecutePolicyAction.java:
```diff
@@ -0,0 +1,170 @@
+
+                @Override
+                public Task createTask(long id, String type, String action, TaskId parentTaskId, Map<String, String> headers) {
+                    String description = "executing enrich policy [" + request.getName() + "]";
```
I did that this way, because overwriting the `getDescription()` method results in more lines of code.

martijnvg commented on x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/action/InternalExecutePolicyAction.java:
```diff
@@ -0,0 +1,170 @@
+                } else {
+                    listener = ActionListener.wrap(result -> {
+                        LOGGER.debug("successfully executed policy [{}]", request.getName());
+                        // taskManager.storeResult(task, new Response(result), ActionListener.wrap(() -> {/*noop*/}));
```
I will remove this. 

I was playing around with storing the result of policy execution in the `.tasks` index. So that we can consistently report on async executed policy executions. Do you think we should do that in a followup change?

jbaiera replied:
I think that's a good idea. I only see a couple of places where we actually store task data after it completes. Seems like Reindex, and a couple of ML actions. I don't see any reason we shouldn't be storing them, though IIRC the user needs to clear them out themselves over time...

martijnvg replied:
Cool, I will add this to #48988

>  though IIRC the user needs to clear them out themselves over time...

True, but I don't expect this index will ever contain a massive amount of data. Ideally this should be a data stream, so that ilm can manage it.

martijnvg replied:
Removed commented out code: 3e455d41ab5f29bdbf70c989aa7b3291eb62735a

martijnvg committed:
In case of wait_for_completion=false policy execution then return the response but not release the policy immediately.
Execute get task request with wait_for_completion=true and when that returns release the policy.

martijnvg committed:
Adjust selectNodeForPolicyExecution() method

martijnvg committed:
remove commented-out code

martijnvg committed:
Merge remote-tracking branch 'es/master' into enrich_delegate_policy_runner_to_non_dedicated_master_node

martijnvg committed:
spotless

jbaiera reviewed (approved):
Thanks @martijnvg, LGTM!
